### PR TITLE
Add RedisSentinel class and tests

### PR DESCRIPTION
**Pull request by yatsukhnenko**
I'm not sure about `getMasterAddrByName`. It returns result as array `[host, port]`, maybe string `host:port` would be better?
TODO: documentation, more tests

### CODE REVIEW

**Comment by yatsukhnenko** [comment]
@michael-grunder don't slow down innovations 😄

**Comment by michael-grunder** [comment]
This works well to communicate with Sentinel but I wonder if it's worth creating a more robust API

I'm just thinking out loud but since Sentinel is a distributed system I wonder if it's worth creating mechanisms where the user could provide multiple Sentinel `host:port` which we could then query in a load balanced way.

Theoretically we could even handle returning the `Redis` object given whichever reply we got from `get-master-addr-by-name`.

Something like:

```php
$arr_sentinels = [['127.0.0.1', 26379], ['127.0.0.1', 26380]];
$obj_redis = RedisSentinel::getMaster($arr_sentinels);
$obj_redis->set('foo', 'bar');
```

I think that might be cool but this interface is working for me now.

**Comment by yatsukhnenko** [comment]
@michael-grunder I like an idea to return `Redis` object in `get-master-addr-by-name`. I'll make the change.

File: sentinel_library.c [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+            redis_sock_set_err(redis_sock, inbuf + 1, len - 1);
+        }
+        RETURN_FALSE;
+    }
+    array_init(&z_ret);
+    nelem = atoi(inbuf + 1);
+    for (i = 0; i < nelem; ++i) {
+        /* redis_mbulk_reply_zipped_raw calls redis_mbulk_reply_zipped
+         * which puts result into return_value via RETVAL_ZVAL */
+        array_init(return_value);
```
[the remark this answers is not in the document]
Comment (yatsukhnenko) [reply sentinel_library.c]: I removed `array_init`. I don't know why but I don't see this leak when run tests locally.

**Comment by guaychou** [comment]
When it will be merged ?

**Comment by themsaid** [comment]
We're dropping support for predis in Laravel due to inactivity, having RedisSentinel support in phpredis would mean everyone using predis can switch to phpredis with no issues.

**Comment by yatsukhnenko** [comment]
I'm going to add some more changes and then merge it after my vacation 🏂

**Comment by guaychou** [comment]
> I'm going to add some more changes and then merge it after my vacation 🏂

Very nice thank you for your contribution

**Comment by yatsukhnenko** [comment]
> I like an idea to return Redis object in `get-master-addr-by-name`

Now I don't :smile: There are two problems:
1. we can return Redis object but it will not be ready to use because there are many other parameters that users usually set via `connect`.
2. method name is `get master address by name` but not `get master by name` so this may confuse users.

@michael-grunder if you don't mind I'm going to add documentation and merge current implementation as is

**Comment by yatsukhnenko** [comment]
@michael-grunder feel free to add any changes that you consider necessary and merge if everything is ok

**Comment by Bro3x** [comment]
Will there be a native sentinel PHP Session handler support? So that you can add sentinels in session.save_path and will have session locking, failover, etc...

**Comment by yatsukhnenko** [comment]
@Bro3x what do you mean?

**Comment by Tommuh** [comment]
He means that the session handler now needs a TCP ip/port where redis lives. By default is doesn't support failover. You need to handle this by your self before setting the IP/port.

**Comment by Bro3x** [comment]
It would be great, if we can handle a common multi node example by default.

I can think about two szenarios. First one has a sentinel on each application node. The session.save_path is that sentinel and the session is stored in the current promoted master.
The second one is more complex and is able to solve a failover if the sentinel on the local node isen't available.

```
   +--------------+         +--------------+
   | application1 |         | application2 |
   +- - - - - - - +         +- - - - - - - +
   |  sentinel4?  |         |  sentinel5?  |
   +--------------+         +--------------+
           |                        |
           |                        |
+------------------sentinels-------------------+
|  +-----------+  +-----------+  +-----------+ |
|  | sentinel1 |  | sentinel2 |  | sentinel3 | |
|  +-----------+  +-----------+  +-----------+ |
+----------------------------------------------+
                       |
                       |
           +---------redis---------+
           |        +----+         |
           |        | M1 |         |
           |        | S1 |         |
           |        +----+         |
           |           |           |
           | +----+    |    +----+ |
           | | R2 |----+----| R3 | |
           | | S2 |         | S3 | |
           | +----+         +----+ |
           +-----------------------+
```

A php.ini could look like
---
session.save_handler = redis
redis.session.locking_enabled = 1
redis.session.lock_expire = 60
redis.session.lock_wait_time = 50000
redis.session.lock_retries = 10

Szenario 1 (application1 only)
---
redis.session.connection_type = sentinel
session.save_path = "tcp://sentinel4:6379?timeout=1.5&persistent=s4&retry_interval=100&read_timeout=1.5"

Szenario 2
---
redis.session.connection_type = sentinel
session.save_path = "tcp://sentinel1:6379?timeout=1.5&persistent=s1&retry_interval=100&read_timeout=1.5,tcp://sentinel2:6379?timeout=1.5&persistent=s2&retry_interval=100&read_timeout=1.5,tcp://sentinel3:6379?timeout=1.5&persistent=s3&retry_interval=100&read_timeout=1.5"

**Comment by yatsukhnenko** [comment]
@Bro3x definitely not in this release. Currently this is basic implementation and we need to get feedback from users and fix bugs after public release. Only then will may think about extending library with features like you described.

**Pull request merged**